### ISSUE TITLE
20190722-20190728

### ISSUE THREAD

**Issue by rogercmq** [rogercmq/NetworkCompressionPaperNotes#1]
### 证明 double masking 在 structured pruning 上 accuracy degradation
目前做的工作: network slimming re-implementation [codes here](https://github.com/rogercmq/NetworkCompressionPaperNotes/tree/master/Network%20Slimming%20Implementation)

- [x] 认真思考一下 [batch_size, output channel num, input feature map H, input feature map W] 在 CONV + BN+ ReLU 上是怎么前向与反向的 

- [ ] 如何实现结构化剪枝的 first mask 就非常重要。应该也没有什么论文可以看了，所以只能每天读一遍 Yuan Xie 的 paper，看能不能从 non-structural 突发奇想到 structural.

- [ ] data-dependent memory 再想一想 binary code 和 decimal float 的关系。。。。。。

- [ ] 先实现 sparse training version 的 Pruning Filters for Efficient ConvNets (ICLR 2017)

- [ ] 加入 double masking 机制之后, 比较 algorithm 与 algorithm + double mask 在三个指标：1. acc；2. 收敛速度，以epoch和iter为单位；3. model稳定性（robustness）上的区别

**Comment by rogercmq** [comment]
### 立的Flag

- [ ] 实现一种 double masking 的算法，主要是确认 first mask 的 metric (目前想到的是 filter pruning by geometric median).  现在的问题：源码有问题。。。。。。

**Comment by rogercmq** [comment]
### network slimming 
不满足项目中“边训边减”的要求。理由：稀疏度提高之后必须 fine-tune, 因为 accuracy 掉得太凶了
### 可能需要复现的文章
可能需要复现的文章：
1. Less is More: Towards compact CNNs —— 稀疏化训练  —— 最早的文章 —— ECCV 2016
2. Compression-aware Training of Deep Networks —— NIPS 2017 —— Less is More 被以下稀疏化引用
3. Learning Structured Sparsity in Deep Neural Networks —— Group LASSO 稀疏化训练

    > add group-sparse regularization on neurons during training to learn compact CNNs with reduced
filters.

4. Learning Structured Sparsity in Deep Learning —— NIPS2016 —— 从题目来看是结构化稀疏

    > add structured sparsity regularizer on each layer to reduce trivial filters, channels or even layers.

5. Fast Convnets Using Group-wise Brain Damage —— CVPR2016 —— 不确定是否为结构化剪枝算法

    > leverage group-sparsity on the convolutional filters to achieve structured brain damage, i.e., prune the entries of the convolution kernel in a group-wise fashion.

6. Learning the number of neurons in deep networks

    > Specifically, our method does not require training an initial network as a pre-processing step …

**Comment by rogercmq** [comment]
Learning to Prune Filters in Convolutional Neural Networks 用到了强化学习，与本项目无关

**Comment by rogercmq** [comment]
network slimming 不满足项目中“边训边减”的要求。理由：稀疏度提高之后必须 fine-tune, 因为 accuracy 掉得太凶了。。。

**Comment by rogercmq** [comment]
### 未开源，如有必要可以复现的文章
1. Play and Prune: Adaptive Filter Pruning for Deep Model Compression
    > Our framework, called Play and Prune (PP), jointly prunes and fine-tunes CNN model parameters, with an adaptive pruning rate, while maintaining the model’s predictive performance.
